plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/analyzer/internal/impl/TableReferencesAnalyzerImpl.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2023 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.model.sql.analyzer.internal.impl;
+
+import org.jkiss.code.NotNull;
+import org.jkiss.code.Nullable;
+import org.jkiss.dbeaver.Log;
+import org.jkiss.dbeaver.model.lsm.LSMAnalyzer;
+import org.jkiss.dbeaver.model.lsm.sql.dialect.LSMDialectRegistry;
+import org.jkiss.dbeaver.model.sql.SQLScriptElement;
+import org.jkiss.dbeaver.model.sql.analyzer.TableReferencesAnalyzer;
+import org.jkiss.dbeaver.model.sql.completion.SQLCompletionRequest;
+import org.jkiss.dbeaver.model.stm.*;
+import org.jkiss.utils.CommonUtils;
+
+import java.io.StringReader;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * Internal implementation of references analyzer
+ */
+public class TableReferencesAnalyzerImpl implements TableReferencesAnalyzer {
+
+    private Map<String, String> tableReferences;
+    private SQLCompletionRequest request;
+    private Log log = Log.getLog(TableReferencesAnalyzerImpl.class);
+
+    public TableReferencesAnalyzerImpl(SQLCompletionRequest request) {
+        this.request = request;
+    }
+
+    private void prepareTableReferences() {
+        final SQLScriptElement activeQuery = request.getActiveQuery();
+        if (activeQuery == null) {
+            tableReferences = Collections.emptyMap();
+            return;
+        }
```

Comment:
Make this method to accept `String query` and move `if (activeQuery == null)` to the caller. Then you can call it from `getTableAlicesFromQuery`. Otherwise we parse the same query multiple times.

Reply:
No objections, good advice.